### PR TITLE
Added mixin support for forge (#155)

### PR DESCRIPTION
**Pull request by MrAn0nym**
(Currently only compatible with fg3)
Additional changes:
 - Modified ForgeProjectCreator.kt to work like FabricProjectCreator.kt)

### CODE REVIEW

**Comment by MrAn0nym** [comment]
image

File: src/main/kotlin/platform/forge/creator/Fg3Template.kt [hunk 1]
```diff
@@ -147,4 +149,16 @@ object Fg3Template : BaseTemplate() {
 
         return project.applyTemplate(PACK_MCMETA_TEMPLATE, props)
     }
+    fun applyMixinConfigTemplate(
```
Comment (RedNesto) [inline src/main/kotlin/platform/forge/creator/Fg3Template.kt:159]: Add a blank line before this one

File: src/main/kotlin/platform/forge/creator/Fg3Template.kt [hunk 2]
```diff
@@ -147,4 +149,16 @@ object Fg3Template : BaseTemplate() {
...
+        project: Project,
+        buildSystem: BuildSystem
+    ): String {
+        val packageName = "${buildSystem.groupId.toPackageName()}.${buildSystem.artifactId.toPackageName()}.mixin"
+        val artifactId = "${buildSystem.artifactId.toPackageName()}"
```
Comment (RedNesto) [inline src/main/kotlin/platform/forge/creator/Fg3Template.kt]: Move this variable up one line and use it in the string interpolation.
Might as well introduce another variable for the groupId to make the whole interpolation more readable.

File: src/main/kotlin/platform/forge/creator/ForgeProjectCreator.kt [hunk 3]
```diff
@@ -54,7 +54,7 @@ class Fg2ProjectCreator(
         val settingsText = Fg2Template.applySettingsGradle(project, buildSystem.artifactId)
         val files = GradleFiles(buildText, propText, settingsText)
 
-        return listOf(
+        val steps = mutableListOf(
```
Comment (RedNesto) [inline src/main/kotlin/platform/forge/creator/ForgeProjectCreator.kt]: Please revert changes to this method.

File: src/main/kotlin/platform/forge/creator/ForgeProjectCreator.kt [hunk 4]
```diff
@@ -69,14 +69,16 @@ class Fg2ProjectCreator(
...
+        return steps
     }
 
     override fun getMultiModuleSteps(projectBaseDir: Path): Iterable<CreatorStep> {
         val buildText = Fg2Template.applySubBuildGradle(project, buildSystem, mcVersion)
         val propText = Fg2Template.applyGradleProp(project, config)
         val files = GradleFiles(buildText, propText, null)
 
-        return listOf(
+        val steps = mutableListOf(
```
Comment (RedNesto) [inline src/main/kotlin/platform/forge/creator/ForgeProjectCreator.kt]: Ditto

File: src/main/kotlin/util/MinecraftTemplates.kt [hunk 5]
```diff
@@ -197,6 +198,7 @@ class MinecraftTemplates : FileTemplateGroupDescriptorFactory {
         const val FORGE_BUILD_GRADLE_TEMPLATE = "Forge build.gradle"
         const val FORGE_SUBMODULE_BUILD_GRADLE_TEMPLATE = "Forge Submodule build.gradle"
         const val FORGE_GRADLE_PROPERTIES_TEMPLATE = "Forge gradle.properties"
+        const val FORGE_MIXINS_JSON_TEMPLATE = "forge_mixins.json"
```
Comment (RedNesto) [inline src/main/kotlin/util/MinecraftTemplates.kt]: Perhaps rename the file to `Forge Mixins Config.json` for consistency?

File: src/main/resources/fileTemplates/j2ee/forge/Forge (1.13+) build.gradle.ft [hunk 6]
```diff
@@ -1,15 +1,18 @@
 buildscript {
     repositories {
         maven { url = 'https://files.minecraftforge.net/maven' }
+        maven { url = 'https://repo.spongepowered.org/repository/maven-public/' }
```
Comment (RedNesto) [inline src/main/resources/fileTemplates/j2ee/forge/Forge (1.13+) build.gradle.ft:5]: Please guard all those additions behind a template condition (similarly to `HAS_DATA` used below in this file), let's not pollute everyone's build.gradle with stuff they won't use.

Also, the refmap configuration is missing.

File: src/main/resources/fileTemplates/j2ee/forge/Forge (1.13+) build.gradle.ft [hunk 7]
```diff
@@ -40,6 +43,8 @@ minecraft {
             // Recommended logging level for the console
             property 'forge.logging.console.level', 'debug'
 
+            arg "-mixin.config=${ARTIFACT_ID}.mixins.json"
```
Comment (RedNesto) [inline src/main/resources/fileTemplates/j2ee/forge/Forge (1.13+) build.gradle.ft:53]: Could you add it to the server configuration too ?

**Review by RedNesto — CHANGES_REQUESTED** [review]
Works correctly. The major issue right now is that the build.gradle unconditionally includes mixins (with the checkbox unchecked in the creator.)

File: src/main/kotlin/platform/forge/creator/Fg3Template.kt (hunk 2 above)
Comment (MrAn0nym) [reply src/main/kotlin/platform/forge/creator/Fg3Template.kt]: >Might as well introduce another variable for the groupId to make the whole interpolation more readable.

Do you mean replacing `${PACKAGE_NAME}` with `${groupId}.${artifactId}`?

Comment (RedNesto) [reply src/main/kotlin/platform/forge/creator/Fg3Template.kt]: I mean
```kotlin
        val groupId = buildSystem.groupId.toPackageName()
        val artifactId = buildSystem.artifactId.toPackageName()
        val packageName = "$groupId.$artifactId.mixin"
```

**Comment by RedNesto** [comment]
Oh, one last thing I forgot: could you make the Use Mixins option enabled/selectable only if the selected Minecraft version is >= 1.16.1 ? (Should be the correct version IIRC)

**Comment by SizableShrimp** [comment]
Mixins are supported on 1.15.2+ and 1.16.1+ dependent on the forge version a player uses, I forget the specific versions. So this is slightly inaccurate.

**Comment by RedNesto** [comment]
You're right, [0.8.2 was added in 1.15.2-31.2.45](https://maven.minecraftforge.net/net/minecraftforge/forge/1.15.2-31.2.45/forge-1.15.2-31.2.45-changelog.txt)

I don't know how much effort would be required to get the mixin version used by a specific forge build. We could use it for the `annotationProcessor` dependency (and determining if the build includes Mixin in the first place).

**Comment by SizableShrimp** [comment]
> 
> 
> You're right, [0.8.2 was added in 1.15.2-31.2.45](https://maven.minecraftforge.net/net/minecraftforge/forge/1.15.2-31.2.45/forge-1.15.2-31.2.45-changelog.txt)
> 
> I don't know how much effort would be required to get the mixin version used by a specific forge build. We could use it for the `annotationProcessor` dependency (and determining if the build includes Mixin in the first place).

`MixinBootstrap.VERSION`

**Comment by RedNesto** [comment]
I doubt we can get it at project creation time ^^. This information might be present in some POM the plugin can download and parse, I can't dig into it for now unfortunately.

File: src/main/kotlin/platform/forge/creator/ForgeProjectSettingsWizard.kt [hunk 8]
```diff
@@ -78,6 +80,13 @@ class ForgeProjectSettingsWizard(private val creator: MinecraftProjectCreator) :
         mcpWarning.isVisible = (mcpVersionBox.selectedItem as? McpVersionEntry)?.isRed == true
     }
 
+    private val forgeVersionBoxListener = ActionListener {
+        val supportedMixinVersion = forgeVersionBox.selectedItem as SemanticVersion >= SemanticVersion.parse("31.2.45")
+        mixinsCheckbox.isEnabled = supportedMixinVersion
+        mixinsCheckbox.isSelected = supportedMixinVersion
+
```
Comment (RedNesto) [inline src/main/kotlin/platform/forge/creator/ForgeProjectSettingsWizard.kt]: The first line is too long, and might fail if `forgeVersionBox.selectedItem` for some reason.
`SemanticVersion.release` can also be used for hardcoded versions:
```suggestion
        val selectedVersion = forgeVersionBox.selectedItem as? SemanticVersion ?: return@ActionListener
        val supportedMixinVersion = selectedVersion >= SemanticVersion.release(31, 2, 45)
        mixinsCheckbox.isEnabled = supportedMixinVersion
        mixinsCheckbox.isSelected = supportedMixinVersion
```

Other than that it looks good enough, the "simple" comparison used here does the job for me, I don't expect people to use forge versions that are more than 6 months old.

**Review by RedNesto — APPROVED** [review]

**Review by RedNesto — CHANGES_REQUESTED** [review]
One last thing (hopefully): could you disable Mixins by default?
We prefer to keep coremodding opt-in for forge.

**Comment by MrAn0nym** [comment]
Done ^^

**Review by RedNesto — APPROVED** [review]

**Pull request merged**